### PR TITLE
OCD-3098: Prevent ONC-ACBs from adding/editing surveillance referencing removed criteria

### PR DESCRIPTION
**Pull request by kekey1**

### CODE REVIEW

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/surveillance/reviewer/SurveillanceDetailsReviewer.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    @Autowired
+    public SurveillanceDetailsReviewer(CertifiedProductDAO cpDao, SurveillanceDAO survDao,
+            ChplProductNumberUtil chplProductNumberUtil, ErrorMessageUtil msgUtil) {
+        this.cpDao = cpDao;
+        this.survDao = survDao;
+        this.chplProductNumberUtil = chplProductNumberUtil;
+        this.msgUtil = msgUtil;
+    }
+
```
Comment (drbgfc) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/surveillance/reviewer/SurveillanceDetailsReviewer.java]: This is just a preference note, not a required change. I'm not sure what we've decided as a team but: I like to see @Override annotations when a method is overridden such as here here, from "void review(Surveillance surveillance)." It helps identify the difference and makes reading/understanding the code faster.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/PendingSurveillanceManager.java [hunk 2]
```diff
@@ -1,31 +1,685 @@
...
+        // to know if we handle it normally or as a background job
+        String data = fileUtils.readFileAsString(file);
+
+        // This is a container used for 2 different result types...
+        SurveillanceUploadResult uploadResult = new SurveillanceUploadResult();
+
+        int numSurveillance = survUploadHelper.countSurveillanceRecords(data);
+        if (numSurveillance < surveillanceThresholdToProcessAsJob) {
+            uploadResult = processAsFile(file);
+        } else { // process as job
```
Comment (drbgfc) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/PendingSurveillanceManager.java]: Normally I'm not the type to advocate against comments, but in this case, the comment is the same as the method name so I think it should probably be removed
// process as job == processUploadAsJob(data);

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/UpdateActionPermissions.java [hunk 3]
```diff
@@ -14,13 +29,25 @@ public boolean hasAccess() {
...
+                }
+            } catch (EntityRetrievalException ex) {
+                LOGGER.error("Could not find listing with ID " + surv.getCertifiedProduct().getId());
+            }
+            if (listing != null && listing.getCertificationBodyId() != null
+                    && isAcbValidForCurrentUser(listing.getCertificationBodyId())) {
+                return true;
+            }
+            return false;
         } else {
```
Comment (drbgfc) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/UpdateActionPermissions.java]: Understood not part of your change, but if I'm reading this correctly, technically, this else is not needed. Returning false alone should suffice.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/AddDocumentActionPermissions.java [hunk 4]
```diff
@@ -20,11 +36,35 @@ public boolean hasAccess(Object obj) {
...
+                return false;
+            }
         } else {
             return false;
         }
     }
 
+    private boolean isNonconformityForRemovedCriteria(SurveillanceEntity surv, Long nonconformityId) {
+        if (!ff4j.check(FeatureList.EFFECTIVE_RULE_DATE_PLUS_ONE_WEEK)) {
+            return false;
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/permissions/domains/surveillance/AddDocumentActionPermissions.java]: I know the ticket is flagged with this flag, but should security care? If there were a removed criteria today we'd still want to restrict ACBs from adding/removing documents related to that criteria, right?

**Pull request merged**